c7n/resources/elasticsearch.py

```diff
@@ -115,6 +117,100 @@ class KmsFilter(KmsRelatedFilter):
     RelatedIdsExpression = 'EncryptionAtRestOptions.KmsKeyId'
 
 
+@ElasticSearchDomain.filter_registry.register('cross-account')
+class ElasticSearchCrossAccountAccessFilter(CrossAccountAccessFilter):
+    """
+    Filter to return all elasticsearch domains with cross account access permissions
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: check-elasticsearch-cross-account
+            resource: aws.elasticsearch
+            filters:
+              - type: cross-account
+    """
+    policy_attribute = 'c7n:Policy'
+    permissions = ('es:DescribeElasticsearchDomainConfig',)
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('es')
+        for r in resources:
+            if self.policy_attribute not in r:
+                result = self.manager.retry(
+                    client.describe_elasticsearch_domain_config,
+                    DomainName=r['DomainName'],
+                    ignore_err_codes=('ResourceNotFoundException',))
+                if result:
+                    r[self.policy_attribute] = json.loads(
+                        result.get('DomainConfig').get('AccessPolicies').get('Options')
+                    )
+        return super().process(resources)
+
+
+@ElasticSearchDomain.action_registry.register('remove-statements')
+class RemovePolicyStatement(RemovePolicyBase):
+    """
+    Action to remove policy statements from elasticsearch
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: elasticsearch-cross-account
+            resource: aws.elasticsearch
+            filters:
+              - type: cross-account
+            actions:
+              - type: remove-statements
+                statement_ids: matched
+    """
+
+    permissions = ('es:DescribeElasticsearchDomainConfig', 'es:UpdateElasticsearchDomainConfig',)
+
+    def validate(self):
+        for f in self.manager.iter_filters():
+            if isinstance(f, ElasticSearchCrossAccountAccessFilter):
+                return self
+        raise PolicyValidationError(
+            '`remove-statements` may only be used in '
+            'conjunction with `cross-account` filter on %s' % (self.manager.data,))
+
+    def process(self, resources):
+        results = []
+        client = local_session(self.manager.session_factory).client('es')
+        for r in resources:
+            try:
+                results += filter(None, [self.process_resource(client, r)])
```

Comment:
nit, this expression feels a bit overly complicated, one simplification
results.extend([self.process_resource()] or [])

separate question though is why return results here, the resource is already annotated with the statements that will be removed.

Reply:
Honestly I was just modeling off of the remove-statements actions I saw in other resources (e.g. in https://github.com/cloud-custodian/cloud-custodian/blob/e07e6ac6b6840a0cb747fa48d0a8ae3954088adb/c7n/resources/glacier.py and https://github.com/cloud-custodian/cloud-custodian/blob/b798824f705ddf606a0be2c92dfad2532cb6bf74/c7n/resources/sqs.py). I've changed it so there's no need for returning results now though.